### DISCUSSION TITLE
Meshroom CUDA alternative

### DISCUSSION THREAD

**Discussion by 1kaiser** [alicevision/Meshroom#1283] [Q&A]
Rcom code for meshroom using hipify code [HIPIFY](https://github.com/ROCm-Developer-Tools/HIPIFY) as nvidia gpu are not generally available these days.

**Comment by carlo-strata** [comment]
This is a very good news! Radeon Open Compute (ROCm)...
https://en.wikipedia.org/wiki/GPUOpen#Radeon_Open_Compute_(ROCm)
https://github.com/RadeonOpenCompute
https://rocmdocs.amd.com/en/latest/
https://www.amd.com/en/graphics/servers-solutions-rocm

nVidia has not always been so addicted to open source... I never understood why! Has Steve Ballmer moved there from Microsoft? ;-) I will check the People that have defined his philosophy in the last 20 years at least... 

I don't understand why meshroom choices to be open source, but use a closed/proprietary technology like CUDA (https://en.wikipedia.org/wiki/CUDA). Is it so fast with respect to OpenCL (https://en.wikipedia.org/wiki/OpenCL), OpenACC (https://en.wikipedia.org/wiki/OpenACC) or even OpenGPU (https://en.wikipedia.org/wiki/GPUOpen ex Heterogeneous-compute Interface for Portability/HIP)???

I will be here to check the news and the open evolution!

I am an Italian Computer Engineer.

Have all a sunny day,

Carlo

**Reply by natowi** [reply]
> I don't understand why meshroom choices to be open source, but use a closed/proprietary technology like CUDA 

CUDA is the de-facto standard in this field. You can read the following links to get an idea on why that is [quora](https://www.quora.com/Why-cant-a-deep-learning-framework-like-TensorFlow-support-all-GPUs-like-a-game-does-Many-games-in-the-market-support-almost-all-GPUs-from-AMD-and-Nvidia-Even-older-GPUs-are-supported-Why-cant-these-frameworks) & [medium](https://medium.com/@stevenn.hansen/what-are-we-actually-doing-on-cuda-and-opencl-1020553bfcdb)
Many operations are easier to implement in CUDA than other GPU frameworks. A HIP port is technically possible, but someone has to do it https://github.com/alicevision/AliceVision/issues/439#issuecomment-783312578

**Reply by carlo-strata** [reply]
A de-facto standard in the Human Beings' Values and of course in the XXI century has to be open and free...
Think, for example, If I want to put my photogrammetry calculi on a number crunching server built on a 4 x 4 x 4 = 64 Raspberry PI 4 (8 Gbyte, 4 or more cores) mini-cluster or something else...

I have to be able to make a driver for the Open API like OpenACC, OpenGPU, ... that the software, meshroom in this case, see as the CPU+GPU or GPU only power.

This is the real Freedom to The Future!

**Reply by natowi** [reply]
I agree - The problem is that all alternatives at the moments a fragile at best. I don´t know of any open source photogrammetry tool that supports computation on non-nvidia GPUs. They all (literally a hand full) use (not performance optimized/ non parallelized) CPU processing. I am testing a CPU based alternative for the DepthMap node, but it is really slow and is only suitable for a two digit number of images. Hopefully this will all change in a few years, once the open GPU alternatives become more complete, better documented and stable.

**Reply by carlo-strata** [reply]
Hi Everyone,

I recently discover this project:
https://github.com/openphotogrammetry/meshroomcl

and consequently this other one:
https://github.com/openphotogrammetry/colmap-cl

Did you know them?

Do you collaborate each other or have been seen only like "enemies"?

I install the first application with success in a AMD card based Windows PC and I run the interface, but not yet some photogrammetry process. I personally use linux all my time and I think that OpenCL has to be considered the first step on a totally free and multi-platform powerful photogrammetry solution.

Have a nice day,

Carlo

p.s. Have this post to be put also as a new item (feature request) to spread the interest about CUDA alternatives?

**Reply by mainmind83** [reply]
I will test MeshroomCL @carlo-strata 


Could [GPUfort](https://github.com/ROCmSoftwarePlatform/gpufort) be used?

**Comment by korenevskiy** [comment]
Please tell me, can the program somehow use AMD Radeon acceleration or just find another free alternative (3DF Zephyr)?
I don't understand anything in these terms. I only know that nVidia uses CUDA and that Mushroom uses CUDA, but I don't understand whether Mushroom can work with Radeon hardware or not?.
I've always liked Radeon, because it can work for a long time without updating drivers. Unlike nVidia, where it was constantly required to update drivers for better performance in modern games. 
And now I have a Radeon video card, I do not know what to do, because there is no CUDA in it. I'm not the only one, there are many of us in the world.

**Reply by natowi** [reply]
Sorry, but no. But you can use draft meshing: https://github.com/alicevision/Meshroom/wiki/Draft-Meshing

Here are two articles which provide some background, why cuda is so dominantly used in general:
[Background](https://www.quora.com/Why-cant-a-deep-learning-framework-like-TensorFlow-support-all-GPUs-like-a-game-does-Many-games-in-the-market-support-almost-all-GPUs-from-AMD-and-Nvidia-Even-older-GPUs-are-supported-Why-cant-these-frameworks) on why CUDA is more popular than OpenCL/Vulkan

[CUDA vs. OpenCL](https://medium.com/@stevenn.hansen/what-are-we-actually-doing-on-cuda-and-opencl-1020553bfcdb)
